core/src/main/scala/kafka/server/FinalizedFeatureCache.scala

```diff
@@ -127,6 +131,22 @@ class FinalizedFeatureCache(private val brokerFeatures: BrokerFeatures) extends
     }
   }
 
+  def update(featuresDelta: FeaturesDelta, highestMetadataOffset: Long): Unit = {
+    val features = featuresAndEpoch.getOrElse(
+      FinalizedFeaturesAndEpoch(Features.emptyFinalizedFeatures(), -1))
+    val newFeatures = new util.HashMap[String, FinalizedVersionRange]()
+    newFeatures.putAll(features.features.features())
+    featuresDelta.changes().entrySet().forEach { e =>
+      e.getValue().asScala match {
+        case None => newFeatures.remove(e.getKey)
+        case Some(feature) => newFeatures.put(e.getKey,
+          new FinalizedVersionRange(feature.min(), feature.max()))
+      }
+    }
+    featuresAndEpoch = Some(FinalizedFeaturesAndEpoch(
+      Features.finalizedFeatures(newFeatures), highestMetadataOffset))
```

Comment:
Wonder if it's worth wrapping `newFeatures` in an unmodifiable map.

Reply:
Yeah, let's do it. This shouldn't change often.